R/applications.R

```diff
@@ -237,3 +237,71 @@ showLogs <- function(appPath = getwd(), appFile = NULL, appName = NULL,
     cat(logs)
   }
 }
+
+#' Sync Application Metadata
+#'
+#' Update the metadata for requested application across all deployments
+#'
+#' @param appPath The path to the directory or file that was deployed.
+#'
+#' @note This function does not update metadata for Shiny and rpubs apps
+#'
+#' @export
+syncAppMetadata <- function(appPath) {
+  if (is.null(appPath) || !file.exists(appPath)) {
+    stop("appPath is null or does not exist")
+  }
+
+  # get all of the currently known deployments
+  deploys <- deployments(appPath)
+
+  syncWindow = getOption("rsconnect.metadata.sync.hours", 24) * 3600
+  now = as.numeric(Sys.time())
+
+  for (i in 1:nrow(deploys)) {
+    lastSyncTime <- deploys[i, "lastSyncTime"]
+
+    # for legacy dcf files that don't have sync time saved yet
+    if (is.null(lastSyncTime) || is.na(lastSyncTime))
+      lastSyncTime <- deploys[i, "when"]
+
+    # don't sync if within the configured time window
+    if (as.numeric(lastSyncTime) + syncWindow > now)
+      next
+
+    # don't sync non-connect apps
+    if (grepl("shinyapps.io", deploys[i, "hostUrl"], fixed = TRUE) ||
+        grepl("rpubs.com", deploys[i, "hostUrl"], fixed = TRUE))
+      next
+
+    account <- rsconnect::accountInfo(deploys[i, "account"],
+                                      server = deploys[i, "server"])
+
+    connect <- clientForAccount(account)
+
+    # if the app does not exist, delete the file
+    tryCatch({
+      application <- connect$getApplication(deploys[i, "appId"])
+
+      # update the record and save out a new config file
+      write.dcf(list(
+        name = deploys[i, "name"],
+        title = application$title,
+        username = deploys[i, "username"],
+        account = deploys[i, "account"],
+        server = deploys[i, "server"],
+        hostUrl = deploys[i, "hostUrl"],
+        appId = deploys[i, "appId"],
+        bundleId = deploys[i, "bundleId"],
+        url = deploys[i, "url"],
+        when = deploys[i, "when"],
+        lastSyncTime = now,
+        asMultiple = deploys[i, "asMultiple"],
+        asStatic = deploys[i, "asStatic"]
+      ), deploys[i, "deploymentFile"])
+    }, error = function(c) {
+      message(paste("appId", deploys[i, "appId"], "no longer exists, deleting config file"))
```

Comment:
Would also recommend that we emit the name and server of the app record we're deleting so that it's human readable ("oh yeah, I did delete that app from that server!")